### PR TITLE
kvserver: factor out replica loading phase

### PR DESCRIPTION
**Pull request by pav-kv**
This commit factors out Replica state loading and invariant checking so that it
is more consolidated rather than interleaved into Replica creation. Replica
creation path was loading this state, to post-factum assert that the in-memory
state matches the state in storage. Now this is a pre-requisite and input into
creating the Replica, and turning it from uninitialized to initialized state.

This change also eliminates the concept of "unloaded" Replica. Now the Replica
is created in a valid state (either initialized or uninitialized).

Touches #93898

### CODE REVIEW

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_init.go:57]: TODO: integrate this with the `stateloader`

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_init.go:103]: TODO: assert `lastIndex` == 0?

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_init.go:107]: TODO: put an `INVARIANT` comment

File: pkg/kv/kvserver/store.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/store.go:1855]: TODO: Here is the integration point with the "global" invariant checking and fixing. Instead of `loadReplica`, we can manually construct `loadedReplicaState` from the batch-loaded set of range descs, etc; then `check` invariants, then `newReplica` manually.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_init.go:243]: Copied this to `newReplica`. It is ok to not leave it here, because the only user of `initRaftMuLockedReplicaMuLocked` is the split trigger, and it rewrites `r.mu.minLeaseProposedTS` right after calling this.

The split trigger essentially does more post-initialization things, which might need to be revisited and included here.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_init.go:294]: This line can probably go away. I didn't like the invariant break conditional to `r.store.TestingKnobs().DisableEagerReplicaRemoval`, but I think it should not be the case on replica creation?

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:294]: Also, `assertStateRaftMuLockedReplicaMuRLocked` can be implemented using `loadReplicaState` now, but I'm not sure what to do about this invariant exception.

**Review by tbg — COMMENTED** [review]
The code LGTM, but I need to understand better where the journey is going to understand if we are moving in the right direction.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:124]: What I had in mind is: `Store.Start` can populate `loadedReplicaState` differently - it knows all the ids/descs etc. But then it would be nice to `check` invariants.

Alternatively, we can do `check` first thing in `newReplica`. This way, `newReplica` is safer (can't accidentally forget to `check`); and there is still flexibility (`loadedReplicaState` can be constructed manually).

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:68]: `assertStateRaftMuLockedReplicaMuRLocked` checks this already.

I think this is already true since #93317?

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:130]: Added a note and TODO.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go]: Done. The check is now hidden in `newReplica` and `initReplica...`.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:190]: Right, I similarly convinced myself that it was ok to lift this, but similarly felt that it's slightly out of place (just like other stuff in this func that is needed only for initialized replicas, e.g. closed timestamp inits etc). Getting the initialization path right should be part of #94912, I think. Maybe I should roll back `newReplica` to always creating an uninited replica first, and then quickly fix it up to become inited.

Orthogonally, this PR is more related to #93898 - it consolidates the storage interaction, so we can tweak it in the scope of sep-raft-log work. This PR, I think, also gives you some more flexibility in `Store.Start`: you can do all the IO there, then create all replicas in memory without hidden assumptions about how replica creation path further interacts with the storage (+reuse the invariant checks).

Comment (tbg) [reply pkg/kv/kvserver/replica_init.go:190]: I'm fine with this PR, especially since both of us independently (didn't see your comment above for some reason where you had very similar reasoning!) arrived at the conclusion that this is OK. I was just trying to see what the next steps would be.

Looks like the plan is to revisit this in #94912 and move "everything" where it belongs, that LGTM. I'll review the new commits now.

Comment (tbg) [reply pkg/kv/kvserver/replica_init.go:190]: Consider leaving a comment on these interim states that explains why it's ok for now and what we're planning to do, etc

File: pkg/kv/kvserver/replica_init.go
Comment (tbg) [reply pkg/kv/kvserver/replica_init.go:68]: Only for initialized replicas, if you have an uninitialized unmigrated replica we'll stumble here, I think.

File: pkg/kv/kvserver/replica_init.go
Comment (tbg) [reply pkg/kv/kvserver/replica_init.go:103]: I think we want some method that asserts integrity of the state, but it should live in kvstorage since it's purely a function of the engine (well, two engines eventually) and this should also live closely to where we do the reconciliation, which again will be somewhere in storage before we even make in-mem replicas.

That said this shouldn't stop us from writing checks. I just think they should at some unspecified point in the future migrate over to `kvstorage` (and also be invoked from `LoadAndReconcileReplicas`).

File: pkg/kv/kvserver/store.go
Comment (tbg) [reply pkg/kv/kvserver/store.go:1855]: Or `LoadAndReconcileReplicas` already checks all of the invariants. This makes sense to me, because the invariants are all on data that's on storage, and by having them in the leaf package they get better testing coverage.

**Review by tbg — COMMENTED** [review]
Fix-up commit looks good. I think we should button this PR up and keep moving, perhaps adding a few more comments about things that are in flux in the code with a direction in which they should move. Mostly to document this in case we get pulled off the case for some reason or other.

I'm pretty interested in seeing the initialization transition cleaned up, since it's extremely brittle and verifying anything takes too much time. Let's discuss early next week.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:68]: Gotcha. And the reason why it didn't break before is that we write `RaftReplicaID` in `getOrCreateReplica` (the only func creating uninited replicas) before calling the invariants check func. With my PR the check moves forward to before creating the `Replica` and can become a problem.

Will wait until your PR merge then.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:68]: Oh wait, actually no. In my PR the load&check phase come after writing `RaftReplicaID` too, so I don't see how this can break.

File: pkg/kv/kvserver/replica_init.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:103]: Agree here, this was my intention too. I'm thinking that `loadedStateReplica` should blend into your `kvstorage.Replica` from #95513 over time, so that loading and invariant checking is all there.

**Comment by pavelkalinnikov** [comment]
@tbg I reverted things a bit, so that this PR is only about factoring out the loading phase. PTAL.

File: pkg/kv/kvserver/store.go
Comment (tbg) [inline pkg/kv/kvserver/store.go:1865]: You're not calling `state.Check()` here, which goes back to my earlier point (can't find the discussion now), should we bake the checks into `Load` so that they can't be skipped?

Comment (tbg) [reply pkg/kv/kvserver/store.go:1865]: I see that `Check` is called inside of `initRaftMuLockedReplicaMuLocked` so it is hit, but still, it's pretty unclear when this method is "supposed" to be called. Calling it automatically at state creation seems like the one reasonable place imo, what's the point of returning an invalid state in the first place?

**Review by tbg — APPROVED** [review]
Nice! The comments don't need to be addressed in this PR.

File: pkg/kv/kvserver/store.go
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/store.go:1865]: Yeah, sounds right, I'll see if `Load` looks better with the embedded check.

My rationale for separating them was related to the fact that we can load a state that doesn't pass the checks, fix it up (e.g. with migrations), and then `Check` before/during passing it to `Replica` constructor. But lemme do a simple thing first, and then revisit if needed.

Comment (pavelkalinnikov) [reply pkg/kv/kvserver/store.go]: Done.

**Comment by pavelkalinnikov** [comment]
bors r=tbg

**Comment by pavelkalinnikov** [comment]
bors retry

**Comment by pavelkalinnikov** [comment]
bors cancel

**Comment by pavelkalinnikov** [comment]
bors retry

**Comment by pavelkalinnikov** [comment]
bors r=tbg

**Pull request merged**
